src/util/asmap.cpp

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) 2019 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <vector>
+#include <assert.h>
+
+int DecodeBits(std::vector<bool> stream, int* bitpos, int minval, const int bit_sizes[], int bit_sizes_n)
+{
+    int val = minval;
+    bool bit;
+    for (int pos = 0; pos < bit_sizes_n; pos++) {
+        int bit_size = bit_sizes[pos];
+        if (pos + 1 < bit_sizes_n) {
+            bit = stream.at(*bitpos);
+            *bitpos = *bitpos + 1;
+        } else {
+            bit = 0;
+        }
+        if (bit) {
+            val += (1 << bit_size);
+        } else {
+            for (int b = 0; b < bit_size; b++) {
+                bit = stream.at(*bitpos);
+                *bitpos = *bitpos + 1;
+                val += bit << (bit_size - 1 - b);
+            }
+            return val;
+        }
+    }
+    return -1;
+}
+
+const int TYPE_BIT_SIZES[] = {0, 0, 1};
+int DecodeType(std::vector<bool> stream, int* bitpos)
+{
+    return DecodeBits(stream, bitpos, 0, TYPE_BIT_SIZES, 3);
+}
+
+const int ASN_BIT_SIZES[] = {15, 16, 17, 18, 19, 20, 21, 22, 23, 24};
+int DecodeASN(std::vector<bool> stream, int* bitpos)
+{
+    return DecodeBits(stream, bitpos, 1, ASN_BIT_SIZES, 10);
+}
+
+const int MATCH_BIT_SIZES[] = {1, 2, 3, 4, 5, 6, 7, 8};
+int DecodeMatch(std::vector<bool> stream, int* bitpos)
+{
+    return DecodeBits(stream, bitpos, 2, MATCH_BIT_SIZES, 8);
+}
+
+const int JUMP_BIT_SIZES[] = {5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30};
+int DecodeJump(std::vector<bool> stream, int* bitpos) {
+    return DecodeBits(stream, bitpos, 17, JUMP_BIT_SIZES, 26);
+}
+
+int BitLength(int target)
+{
+    int result = 0;
+    while (target > 0) {
+        target = target / 2;
+        result++;
+    }
+    return result;
+}
+
+int Interpret(std::vector<bool> asmap, unsigned long long num, int bits)
```

Comment:
Don't pass the map by value; it's copying the entire megabyte bitmap on every call.